lib/Dialects/VectorExt/Transform/VectorWarpUtils.cpp

```diff
@@ -0,0 +1,261 @@
+#include "Dialects/VectorExt/VectorExtWarpUtils.h"
+#include "mlir/Dialect/Affine/IR/AffineOps.h"
+#include "mlir/Dialect/Vector/VectorOps.h"
+#include "mlir/IR/AffineMap.h"
+#include "mlir/IR/BlockAndValueMapping.h"
+#include "mlir/Support/LogicalResult.h"
+
+using namespace mlir;
+using namespace mlir::vector;
+using namespace mlir::vector_ext;
+
+// Clones `op` into a new operations that takes `operands` and returns
+// `resultTypes`.
+static Operation *cloneOpWithOperandsAndTypes(OpBuilder &builder, Location loc,
+                                              Operation *op,
+                                              ArrayRef<Value> operands,
+                                              ArrayRef<Type> resultTypes) {
+  OperationState res(loc, op->getName().getStringRef(), operands, resultTypes,
+                     op->getAttrs());
+  return builder.createOperation(res);
+}
+
+/// Helper to create a new WarpSingleLaneOp regions with different signature.
+static WarpSingleLaneOp
+moveRegionToNewWarpSignature(OpBuilder &b, WarpSingleLaneOp warpOp,
+                             ValueRange newYieldedValues,
+                             TypeRange newReturnTypes) {
+  // Create a new op before the existing one, with the extra operands.
+  OpBuilder::InsertionGuard g(b);
+  b.setInsertionPoint(warpOp);
+  auto newWarpOp = b.create<WarpSingleLaneOp>(warpOp.getLoc(), newReturnTypes,
+                                              warpOp.laneid());
+
+  Region &opBody = warpOp.getBodyRegion();
+  Region &newOpBody = newWarpOp.getBodyRegion();
+  newOpBody.takeBody(opBody);
+  auto yield =
+      cast<vector_ext::YieldOp>(newOpBody.getBlocks().begin()->getTerminator());
+  yield.operandsMutable().assign(newYieldedValues);
+  return newWarpOp;
+}
+
+/// Helper to create a new WarpSingleLaneOp region with extra outputs.
+static WarpSingleLaneOp
+moveRegionToNewWarpAddReturn(OpBuilder &b, WarpSingleLaneOp warpOp,
+                             ValueRange newYieldedValues,
+                             TypeRange newReturnTypes) {
+
+  SmallVector<Type> types(warpOp.getResultTypes().begin(),
+                          warpOp.getResultTypes().end());
+  types.append(newReturnTypes.begin(), newReturnTypes.end());
+  auto yield = cast<vector_ext::YieldOp>(
+      warpOp.getBodyRegion().getBlocks().begin()->getTerminator());
+  SmallVector<Value> yieldValue(yield.getOperands().begin(),
+                                yield.getOperands().end());
+  yieldValue.append(newYieldedValues.begin(), newYieldedValues.end());
+  WarpSingleLaneOp newWarpOp =
+      moveRegionToNewWarpSignature(b, warpOp, yieldValue, types);
+  for (auto it :
+       llvm::zip(warpOp.getResults(),
+                 newWarpOp.getResults().take_front(warpOp.getNumResults())))
+    std::get<0>(it).replaceAllUsesWith(std::get<1>(it));
+  return newWarpOp;
+}
+
+llvm::Optional<std::pair<Operation *, unsigned>>
+getWarpResult(WarpSingleLaneOp warpOp, std::function<bool(Operation *)> fn) {
```

Comment:
Seems like this should be `OpOperand *getWarpYieldOperand` ?

You can then easily get to the defining op, the result number, operand number etc from just that?

Reply:
Yes I was considering doing that but OpOperand get invalidated once I change the yield operand so this avoid potential mistake of using OpOperand after it being killed.

Reply:
well that sounds like a red herring; if you need to memoize, you'd just:
```
Operation *op = yieldOperand->get().getDefiningOp();
int64_t operandIndex = yieldOperand->getOperandNumber();
```
and then not use `yieldOperand`.

It is bad form to be creative with types and APIs for things that are well-supported.

All the `yieldOperand->first` and `yieldOperand->second` should disappear.

Reply:
That makes sense. I changed the signature as suggested.